### PR TITLE
Backup and restore fixes

### PR DESCRIPTION
**Pull request by marceloneppel**
## Issue
[DPE-1584](https://warthogs.atlassian.net/browse/DPE-1584) was reported for this charm (issue with slash not present in the S3 path).

https://github.com/canonical/postgresql-operator/issues/97, https://github.com/canonical/postgresql-operator/issues/99 and https://github.com/canonical/postgresql-operator/issues/129 were reported for the VM charm, but also affect this charm.

## Solution
This PR fixes [DPE-1584](https://warthogs.atlassian.net/browse/DPE-1584) and also fixes in this charm the issues reported for the VM charm on [DPE-1624](https://warthogs.atlassian.net/browse/DPE-1624) and [DPE-1625](https://warthogs.atlassian.net/browse/DPE-1625).

It's a replacement for https://github.com/canonical/postgresql-k8s-operator/pull/143, after we agreed that only one relation is needed for both backups and restores.

* Enabled the restore in another cluster by using a `restore_stanza` relation variable, which is used in the restore command to restore the backup from the other cluster stanza.

* Created a compound stanza name, by adding the model name along with the cluster name.

* Alerted the user with a blocked status if trying to configure backups to a S3 path where there are already backups from another cluster (`can_use_s3_repository ` method - which is called when S3 settings change and also after a restore).

* Unblock the charm when new S3 settings are informed (and are valid). This was done through checking the blocked status messages.

* Added `/` to the beginning of the S3 path when it's not configured with that in the S3 integrator.

* If the bucket provided through S3 integrator doesn't exist, create it in the charm.

New integration tests were added to ensure the backups can be restored in a different cluster (after relating to the S3 integrator, a blocked status will be set, because the backups are from another cluster, but the restore action can be called - later the user can change the S3 path to remove the blocked status) and to ensure the cluster recovers from blocked statuses related to wrong S3 settings.

[DPE-1624]: https://warthogs.atlassian.net/browse/DPE-1624?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DPE-1625]: https://warthogs.atlassian.net/browse/DPE-1625?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DPE-1584]: https://warthogs.atlassian.net/browse/DPE-1584?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/backups.py [hunk 1]
```diff
@@ -326,7 +418,7 @@ def _on_create_backup_action(self, event) -> None:
                 # on the replicas (that happens when TLS is not enabled).
                 command.append("--no-backup-standby")
             stdout, stderr = self._execute_command(command)
-            backup_id = self._list_backups_ids(show_failed=True)[-1]
+            backup_id = list(self._list_backups(show_failed=True).keys())[-1]
```
Comment (delgod) [inline src/backups.py:421]: `.keys()` returns keys in random order, we need or sort it or use OrderedDict in `_list_backups`

**Review by delgod — CHANGES_REQUESTED** [review]

File: src/backups.py [hunk 2]
```diff
@@ -194,26 +277,28 @@ def _initialise_stanza(self) -> None:
...
             logger.exception(e)
-            self.charm.unit.status = BlockedStatus(
-                f"failed to initialize stanza with error {str(e)}"
-            )
+            self.charm.unit.status = BlockedStatus(FAILED_TO_INITIALIZE_STANZA_ERROR_MESSAGE)
             return
 
         # Store the stanza name to be used in configurations updates.
-        self.charm.app_peer_data.update({"stanza": self.charm.cluster_name})
+        self.charm.app_peer_data.update({"stanza": self.stanza_name})
```
[the remark this answers is not in the document]
Comment (marceloneppel) [reply src/backups.py:304]: Great catch! I added the logic to remove the stanza on https://github.com/canonical/postgresql-k8s-operator/pull/147/commits/0bbe4d0d6674b6e57967da39c8f7d9256dc4ff69 if `pgbackrest check` fails. I kept the stanza name here as it's needed to be used in the archive command, which is what should succeed when `pgbackrest check` is called.

File: src/backups.py (hunk 1 above)
Comment (marceloneppel) [reply src/backups.py:428]: Fixed on https://github.com/canonical/postgresql-k8s-operator/pull/147/commits/177f19023274bd3ea2659f1aa664293b568200f4 by using `OrderedDict`.

File: src/backups.py [hunk 3]
```diff
@@ -108,6 +139,45 @@ def _construct_endpoint(self, s3_parameters: Dict) -> str:
...
+
+        bucket_name = s3_parameters["bucket"]
+        region = s3_parameters.get("region")
+        session = boto3.session.Session(
+            aws_access_key_id=s3_parameters["access-key"],
+            aws_secret_access_key=s3_parameters["secret-key"],
+            region_name=s3_parameters["region"],
+        )
+
+        logger.error(f"s3_parameters: {s3_parameters}")
```
[the remark this answers is not in the document]
Comment (marceloneppel) [reply src/backups.py]: My bad. It was a leftover from tests that I was performing locally. Removed on https://github.com/canonical/postgresql-k8s-operator/pull/147/commits/28dd5dbba3f52c1ef0d29baff37737ce301633d4.

**Review by taurus-forever — APPROVED** [review]
Superb, as always! Looking forward to test it from edge channel!

**Pull request merged**